design_docs/crf.md

```diff
@@ -0,0 +1,153 @@
+TODO: add a PoC code for this
+
+This a workaround solution for loading CRF model from disk.
+
+1. Load the model without compile
+```python
+new_model = keras.models.load_model('path_to_my_model.h5', compile=Flase)
+```
+
+2. Get the CRF layer instance
+```python
+# normally, crf layer is the last layer
+crf_layer_instance = new_model.get_layer(index=-1)
+```
+
+3. Get the CRF loss instance from layer instance
+```python
+crf_loss_instance = crf_layer_instance.get_keras_loss()
+```
+
+4. Compile the model
+```python
+new_model.compile(loss=crf_loss_instance)
+```
+
+The shortage of this method is user need to add extract code to load the model and all the arguments except the loss passed to model's compile method before will not longer remembered, user need to pass to it again (if their still remember it)
+
+## About CRF loss
+
+### Solution 1: inherit from tf.keras.losses.Loss
+
+#### pros
+the recommended way to implement a "normal" loss
+
+#### cons
+according to the code around `tensorflow_core/python/keras/engine/training.py:1651`
+`per_sample_losses` returned by `loss_fn.call(y_true, y_pred)` must (or can be converted to) have the same shape with `sample_weight` which default to output `mask` (tensorflow_core/python/keras/engine/training.py:1642) of CRF layer.
+
+but that is not possible because `per_sample_losses` is a 1d tensor and `mask` of CRF is a 2d tensor.
+
+One way to fix it is set output `mark` of crf layer to a 1d tensor, which make the mark is considered as not the same meaning as it's name.
+
+Other way is modified the output of loss class to make `per_sample_losses` to a 2d tensor and properly set the reduce property of the class. It so wired and break the semantic meaning of the interface, should considered to a bad idea.
+
+
+### Solution 2: implement as a function ###
+
+#### pros ####
+This is a old but standard (keras style) way to implement the loss function
+
+#### cons ####
+TensorFlow will convert a loss function into a subclass of `tf.keras.losses.Loss` in `` file by `` (call chain: `tf.keras.Model::compile()` [Line: 314] > `tensorflow/python/keras/engine/training_utils.py::prepare_loss_functions` [Line: 1501] > `tensorflow/python/keras/engine/training_utils.py::get_loss_function` [Line: 1186]).
+
+```python
+  # For losses which are given as strings/functions in the compile API,
+  # we always set the loss reduction type to be `SUM_OVER_BATCH_SIZE`
+  # (both in distribution strategy context and otherwise).
+  return losses.LossFunctionWrapper(
+      loss_fn,
+      name=loss_fn.__name__,
+      reduction=losses_utils.ReductionV2.SUM_OVER_BATCH_SIZE)
+```
+
+So it has same issue that solution 1.
+
+### Solution 3: implement loss as a callable class
+
+#### pros
+Nothing breaks. `mark` property is still a meaningful tensor which standard as a mark.
+
+#### cons
+this solution need understanding how keras process a loss function, which is not documented and not recommend way in TF 2.x.
```

Comment:
I'm not sure that's true anymore: https://www.tensorflow.org/api_docs/python/tf/keras/losses/Loss

I don't fully understand the complexity of CRF but I'll try to make this pull request progress.